From 433de67f5910d780a3b26b19efa05ec3fd86478e Mon Sep 17 00:00:00 2001
From: Shuai Zhang <quic_shuaz@quicinc.com>
Date: Wed, 11 Jun 2025 17:08:46 +0800
Subject: [PATCH v1] FORMLIST: driver: bluetooth: hci_qca: disable hci_auto_off
 when no bt_en

If the BT SoC BT_EN is controlled by hardware, disable the AUTO_OFF
feature. Otherwise, BT will close the HCI layer except for the UART
after firmware download. However, the SoC remains active. If the SoC
sends a packet to the Host after firmware download, the Host cannot
respond since the HCI layer is closed, which will cause the firmware
to enter an incorrect state.

Signed-off-by: Shuai Zhang <quic_shuaz@quicinc.com>

Upstream-Status: Submitted [https://patchwork.kernel.org/project/bluetooth/patch/20250609105538.4090716-1-quic_shuaz@quicinc.com/]

Signed-off-by: Shuai Zhang <quic_shuaz@quicinc.com>
---
 drivers/bluetooth/hci_qca.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/bluetooth/hci_qca.c b/drivers/bluetooth/hci_qca.c
index b9d0bfb68f18..ea673cd942cc 100644
--- a/drivers/bluetooth/hci_qca.c
+++ b/drivers/bluetooth/hci_qca.c
@@ -2476,6 +2476,16 @@ static int qca_serdev_probe(struct serdev_device *serdev)
 	if (power_ctrl_enabled) {
 		set_bit(HCI_QUIRK_NON_PERSISTENT_SETUP, &hdev->quirks);
 		hdev->shutdown = qca_power_off;
+	} else {
+	/*
+	 * If the BT SoC BT_EN is controlled by hardware, disable the AUTO_OFF
+	 * feature. Otherwise, BT will close the HCI layer except for
+	 * the UART after firmware download. However, the SoC remains active.
+	 * If the SoC sends a packet to the Host after firmware download,
+	 * the Host cannot respond since the HCI layer is closed, which will
+	 * cause the firmware to enter an incorrect state.
+	 */
+		hci_dev_clear_flag(hdev, HCI_AUTO_OFF);
 	}
 
 	if (data) {
-- 
2.34.1

